src/number_theory/well_approximable.lean

```diff
@@ -0,0 +1,313 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import dynamics.ergodic.add_circle
+import measure_theory.covering.liminf_limsup
+import data.nat.totient
+
+/-!
+# Well-approximable numbers and Gallagher's ergodic theorem
+
+Gallagher's ergodic theorem is a result in metric number theory. It thus belongs to that branch of
+mathematics concerning arithmetic properties of real numbers which hold almost eveywhere with
+respect to the Lebesgue measure.
+
+Gallagher's theorem concerns the approximation of real numbers by rational numbers. The input is a
+sequence of distances `δ₁, δ₂, ...`, and the theorem concerns the set of real numbers `x` for which
+there is an infinity of solutions to:
+$$
+  |x - m/n| < δₙ,
+$$
+where the rational number `m/n` is in lowest terms. The result is that for any `δ`, this set is
+either almost all `x` or almost no `x`.
+
+This result was proved by Gallagher in 1959
+[P. Gallagher, *Approximation by reduced fractions*](Gallagher1961). It is formalised here as
+`add_circle.add_well_approximable_ae_empty_or_univ` except with `x` belonging to the circle `ℝ ⧸ ℤ`
+since this turns out to be more natural.
+
+Given a particular `δ`, the Duffin-Schaeffer conjecture (now a theorem) gives a criterion for
+deciding which of the two cases in the conclusion of Gallagher's theorem actually occurs. It was
+proved by Koukoulopoulos and Maynard in 2019
+[D. Koukoulopoulos, J. Maynard, *On the Duffin-Schaeffer conjecture*](KoukoulopoulosMaynard2020).
+
+## Main definitions and results:
+
+ * `approx_order_of`: in a seminormed group `A`, given `n : ℕ` and `δ : ℝ`, `approx_order_of A n δ`
+   is the set of elements within a distance `δ` of a point of order `n`.
+ * `well_approximable`: in a seminormed group `A`, given a sequence of distances `δ₁, δ₂, ...`,
+   `well_approximable A δ` is the limsup as `n → ∞` of the sets `approx_order_of A n δₙ`.
```

Comment:
It would be nice to say a bit more explicitly what is the limsup of a sequence of sets, as you do in the relevant docstring.